src/project.rs

```diff
@@ -135,43 +148,104 @@ impl Project {
         }
     }
 
-    pub fn load_from_slice(
+    /// Sets the name of a project. The order it handles is as follows:
+    ///
+    /// - If the project is a `default.project.json`, uses the folder's name
+    /// - If a fallback is specified, uses that blindly
+    /// - Otherwise, loops through sync rules (including the default ones!) and
+    ///   uses the name of the first one that matches and is a project file
+    fn set_file_name(&mut self, fallback: Option<&str>) -> Result<(), Error> {
+        let file_name = self
+            .file_location
+            .file_name()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| Error::ProjectNameInvalid {
+                path: self.file_location.clone(),
+            })?;
+
+        // If you're editing this to be generic, make sure you also alter the
+        // snapshot middleware to support generic init paths.
+        if file_name == PROJECT_FILENAME {
+            let folder_name = self.folder_location().file_name().and_then(OsStr::to_str);
+            if let Some(folder_name) = folder_name {
+                self.name = Some(folder_name.to_string());
+            } else {
+                return Err(Error::FolderNameInvalid {
+                    path: self.file_location.clone(),
+                });
+            }
+        } else if let Some(fallback) = fallback {
+            self.name = Some(fallback.to_string());
+        } else {
+            for rule in self.sync_rules.iter().chain(default_sync_rules()) {
+                // I don't believe there's any way to get here without the
+                // matching middleware being a Project middleware.
+                if rule.matches(&self.file_location) {
+                    self.name = Some(
+                        rule.file_name_for_path(&self.file_location)
+                            .map_err(|_| Error::ProjectNameInvalid {
+                                path: self.file_location.clone(),
+                            })?
+                            .to_string(),
+                    )
+                }
+            }
+            return Err(Error::ProjectNameInvalid {
+                path: std::mem::take(&mut self.file_location),
+            });
```

Comment:
This code path is not being tested - I'm not sure if it's actually reachable? I think the only way we could possibly reach this right now is through the `load_exact` call in ServeSession, which is the only place where we pass `None` for the fallback name - but since serve session cannot load a project that doesn't end in `.project.json`, I don't think sync rules can apply?

Reply:
I don't think it's reachable right now, but it's something I want to handle now instead of just adding a `todo!` or something similar. That said, I can write a test for it if you'd like.

Reply:
Yeah, try to find a way to exercise it in a test - but if it's impossible to test without changes to ServeSession's root project resolution, I prefer a `todo!` (also as it stands, there's no early return once we find a matching middleware, so this code path will always error)

I also question whether it will ever make sense to allow sync rules to override ServeSession's behavior here - maybe the fallback name should just not be optional?